src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java

```diff
@@ -1700,261 +1573,414 @@ protected void move(File source, File target, String entity) throws IOException,
+
+        // Get pid and cid refs files
+        File pidRefsTmpFile = writeRefsFile(cid, HashStoreIdTypes.pid.getName());
+        File cidRefsTmpFile = writeRefsFile(pid, HashStoreIdTypes.cid.getName());
+        // Move refs files to permanent location
+        File absPathPidRefsFile = absPidRefsPath.toFile();
+        File absPathCidRefsFile = absCidRefsPath.toFile();
+        move(pidRefsTmpFile, absPathPidRefsFile, "refs");
+        move(cidRefsTmpFile, absPathCidRefsFile, "refs");
+        // Verify tagging process, this throws an exception if there's an issue
+        verifyHashStoreRefsFiles(pid, cid, absPidRefsPath, absCidRefsPath);
+        logFileHashStore.info(
+            "Object with cid: " + cid + " has been tagged successfully with pid: " + pid);
+    }
+
+    /**
+     * Untags a data object in HashStore by deleting the 'pid reference file' and removing the 'pid'
+     * from the 'cid reference file'. This method will never delete a data object.
+     *
+     * @param pid Persistent or authority-based identifier
+     * @param cid Content identifier of data object
+     * @throws InterruptedException     When there is a synchronization issue
+     * @throws NoSuchAlgorithmException When there is an algorithm used that is not supported
+     * @throws IOException              When there is an issue deleting refs files
+     */
+    protected void unTagObject(String pid, String cid) throws InterruptedException,
+        NoSuchAlgorithmException, IOException {
+        // Validate input parameters
+        FileHashStoreUtility.ensureNotNull(pid, "pid", "unTagObject");
+        FileHashStoreUtility.checkForEmptyAndValidString(pid, "pid", "unTagObject");
+        FileHashStoreUtility.ensureNotNull(cid, "cid", "unTagObject");
+        FileHashStoreUtility.checkForEmptyAndValidString(cid, "cid", "unTagObject");
+
+        Collection<Path> deleteList = new ArrayList<>();
+
+        try {
+            synchronizeObjectLockedPids(pid);
+            // Before we begin untagging process, we look for the `cid` by calling
+            // `findObject` which will throw custom exceptions if there is an issue with
+            // the reference files, which help us determine the path to proceed with.
+            try {
+                Map<String, String> objInfoMap = findObject(pid);
+                cid = objInfoMap.get("cid");
+                // If no exceptions are thrown, we proceed to synchronization based on the `cid`
+                synchronizeObjectLockedCids(cid);
+
+                try {
+                    // Get paths to reference files to work on
+                    Path absCidRefsPath = getHashStoreRefsPath(cid, HashStoreIdTypes.cid.getName());
+                    Path absPidRefsPath = getHashStoreRefsPath(pid, HashStoreIdTypes.pid.getName());
+
+                    // Begin deletion process
+                    updateRefsFile(pid, absCidRefsPath, "remove");
+                    if (Files.size(absCidRefsPath) == 0) {
```

Comment:
If we call Files.size on a data object that does not exist, it will throw a `NoSuchFileException` (tested this).

Note, before the code can ever reach this point - we check that the files are where they expect to be through `findObject`. If this "check" passes, we immediately synchronize/lock the `cid`.